disperser/apiserver/server.go

```diff
@@ -303,6 +304,107 @@ func (s *DispersalServer) disperseBlob(ctx context.Context, blob *core.Blob, aut
 	}, nil
 }
 
+func (s *DispersalServer) DispersePaidBlob(ctx context.Context, req *pb.DispersePaidBlobRequest) (*pb.DisperseBlobReply, error) {
+	blob, err := s.validatePaidRequestAndGetBlob(ctx, req)
+	quorumNumbers := req.CustomQuorumNumbers
+	binIndex := req.PaymentHeader.BinIndex
+	cumulativePayment := new(big.Int).SetBytes(req.PaymentHeader.CumulativePayment)
+	//todo: before disperse blob, validate the signature
+	signature := req.PaymentSignature
+
+	if err != nil {
+		for _, quorumID := range req.CustomQuorumNumbers {
+			s.metrics.HandleFailedRequest(codes.InvalidArgument.String(), fmt.Sprint(quorumID), len(req.GetData()), "DispersePaidBlob")
+		}
+		s.metrics.HandleInvalidArgRpcRequest("DispersePaidBlob")
+		return nil, api.NewInvalidArgError(err.Error())
+	}
+
+	reply, err := s.dispersePaidBlob(ctx, blob, quorumNumbers, binIndex, cumulativePayment, signature, "", "DispersePaidBlob")
+	if err != nil {
+		// Note the DispersePaidBlob already updated metrics for this error.
+		s.logger.Info("failed to disperse blob", "err", err)
+	} else {
+		s.metrics.HandleSuccessfulRpcRequest("DispersePaidBlob")
+	}
+	return reply, err
+}
+
+// dispersePaidBlob checks for payment metering, otherwise does the same thing as disperseBlob
+func (s *DispersalServer) dispersePaidBlob(ctx context.Context, blob *core.Blob, quorumNumbers []uint32, binIndex uint32, cumulativePayment *big.Int, signature []byte, authenticatedAddress string, apiMethodName string) (*pb.DisperseBlobReply, error) {
+	timer := prometheus.NewTimer(prometheus.ObserverFunc(func(f float64) {
+		s.metrics.ObserveLatency("DispersePaidBlob", f*1000) // make milliseconds
+	}))
+	defer timer.ObserveDuration()
+
+	securityParams := blob.RequestHeader.SecurityParams
+	securityParamsStrings := make([]string, len(securityParams))
+	for i, sp := range securityParams {
+		securityParamsStrings[i] = sp.String()
+	}
+
+	blobSize := len(blob.Data)
+
+	origin, err := common.GetClientAddress(ctx, s.rateConfig.ClientIPHeader, 2, true)
+	if err != nil {
+		for _, param := range securityParams {
+			s.metrics.HandleFailedRequest(codes.InvalidArgument.String(), fmt.Sprintf("%d", param.QuorumID), blobSize, apiMethodName)
+		}
+		s.metrics.HandleInvalidArgRpcRequest(apiMethodName)
+		return nil, api.NewInvalidArgError(err.Error())
+	}
+
+	s.logger.Debug("received a new paid blob dispersal request", "authenticatedAddress", authenticatedAddress, "origin", origin, "blobSizeBytes", blobSize, "securityParams", strings.Join(securityParamsStrings, ", "))
+
+	// payments before ratelimits
+	if s.meterer != nil {
```

Comment:
How does it interact with rate limiter? Would there ever be a case where request is limited by both meterer and rate limiter? i.e. is it possible that there's been sufficient amount paid, but is above the rate limit?

Reply:
There won't ever be a case where request is limited by both meterer and rate limiter. If meterer is configured, rate limiter should not be used. 
Do we want a different type of behavior? Like, using meterer if there's `PaymentMetadata`, rate limiter if not?

Reply:
What about something simple like this?
- If the payment metadata is included, we use the meterer. If the meterer is unavailable, we fail. 
- If the payment metadata is not included, we use the rate limiter. If rate limiter is unavailable, pass (current behavior)

The DisperseBlobPaid method will always pass the payment metadata.

Reply:
great, now `disperseBlob` will check for payment metadata to use metering, and then ratelimiter if it exists, otherwise continues as usual